src/Storages/StorageMerge.cpp

```diff
@@ -604,142 +605,164 @@ std::vector<ReadFromMerge::ChildPlan> ReadFromMerge::createChildrenPlans(SelectQ
+            size_t current_need_streams = tables_count >= num_streams ? 1 : (num_streams / tables_count);
+            size_t current_streams = std::min(current_need_streams, remaining_streams);
+            remaining_streams -= current_streams;
+            current_streams = std::max(1uz, current_streams);
 
-        Names column_names_as_aliases;
-        Names real_column_names = column_names;
-
-        const auto & database_name = std::get<0>(table);
-        const auto & table_name = std::get<3>(table);
-        auto row_policy_filter_ptr = modified_context->getRowPolicyFilter(
-            database_name,
-            table_name,
-            RowPolicyFilterType::SELECT_FILTER);
-        if (row_policy_filter_ptr && !row_policy_filter_ptr->empty())
-        {
-            row_policy_data_opt = RowPolicyData(row_policy_filter_ptr, storage, modified_context);
-            row_policy_data_opt->extendNames(real_column_names);
-        }
+            bool sampling_requested = query_info.query->as<ASTSelectQuery>()->sampleSize() != nullptr;
+            if (query_info.table_expression_modifiers)
+                sampling_requested = query_info.table_expression_modifiers->hasSampleSizeRatio();
 
-        auto modified_query_info
-            = getModifiedQueryInfo(modified_context, table, nested_storage_snapshot, real_column_names, column_names_as_aliases, aliases);
+            /// If sampling requested, then check that table supports it.
+            if (sampling_requested && !storage->supportsSampling())
+                throw Exception(ErrorCodes::SAMPLING_NOT_SUPPORTED, "Illegal SAMPLE: table {} doesn't support sampling", storage->getStorageID().getNameForLogs());
 
-        if (!context->getSettingsRef()[Setting::allow_experimental_analyzer])
-        {
-            auto storage_columns = storage_metadata_snapshot->getColumns();
-            auto syntax_result = TreeRewriter(context).analyzeSelect(
-                modified_query_info.query, TreeRewriterResult({}, storage, nested_storage_snapshot));
+            Aliases aliases;
+            RowPolicyDataOpt row_policy_data_opt;
+            auto storage_metadata_snapshot = storage->getInMemoryMetadataPtr();
 
-            bool with_aliases = common_processed_stage == QueryProcessingStage::FetchColumns && !storage_columns.getAliases().empty();
-            if (with_aliases)
+            if (storage_metadata_snapshot->getColumns().empty())
             {
-                ASTPtr required_columns_expr_list = std::make_shared<ASTExpressionList>();
-                ASTPtr column_expr;
+                /// (Assuming that view has empty list of columns iff it's parametrized.)
+                if (storage->isView())
+                    throw Exception(ErrorCodes::STORAGE_REQUIRES_PARAMETER, "Parametrized view can't be queried through a Merge table.");
+                else
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Table has no columns.");
```

Comment:
(The real changes are: this, `getLogger` call above, and try-catch outside this. Everything else is indentation change from wrapping code in try-catch. Idk why git/github's diff algorithm failed so badly on this.)